pkg/controller/rbac/serviceprincipal/managed.go

```diff
@@ -0,0 +1,158 @@
+	errCreateServicePrincipal = "cannot create ServicePrincipal"
+	errUpdateServicePrincipal = "update ServicePrincipal"
+	errGetServicePrincipal    = "cannot get ServicePrincipal"
+	errDeleteServicePrincipal = "cannot delete ServicePrincipal"
+)
+
+// Setup adds a controller that reconciles ServicePrincipal.
+func Setup(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1alpha1.ServicePrincipalKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.ServicePrincipal{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.ServicePrincipalGroupVersionKind),
+			// Override default initializers in case to remove NewNameAsExternalName Initializer
+			managed.WithInitializers(),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{client: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	client client.Client
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	creds, _, err := azureclients.GetAuthInfo(ctx, c.client, mg)
+	if err != nil {
+		return nil, err
+	}
+	cfg, err := adal.NewOAuthConfig(creds[azure.CredentialsKeyActiveDirectoryEndpointURL], creds[azure.CredentialsKeyTenantID])
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create OAuth configuration")
+	}
+	token, err := adal.NewServicePrincipalToken(*cfg,
+		creds[azure.CredentialsKeyClientID],
+		creds[azure.CredentialsKeyClientSecret],
+		creds[azure.CredentialsKeyActiveDirectoryGraphResourceID])
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create service principal token")
+	}
+	if err := token.Refresh(); err != nil {
+		return nil, errors.Wrap(err, "cannot refresh service principal token")
+	}
+	ta := autorest.NewBearerAuthorizer(token)
+	spc := graphrbac.NewServicePrincipalsClient(creds[azure.CredentialsKeyTenantID])
+	spc.Authorizer = ta
+	return &external{spc: spc}, nil
+}
+
+type external struct {
+	spc graphrbacapi.ServicePrincipalsClientAPI
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	s, ok := mg.(*v1alpha1.ServicePrincipal)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotServicePrincipal)
+	}
+	if meta.GetExternalName(s) == "" {
+		return managed.ExternalObservation{}, nil
+	}
+	_, err := e.spc.Get(ctx, meta.GetExternalName(s))
+	if err != nil {
+		return managed.ExternalObservation{}, resource.IgnoreNotFound(errors.Wrap(err, errGetServicePrincipal))
```

Comment:
Please see the comment on error handling in Application's `external.Observe`. I think we should be doing the not-found-error check via `azureclients.IsNotFound`.